extensions/amp-story/1.0/bookend/components/cta-link.js

```diff
@@ -0,0 +1,115 @@
+/**
+ * Copyright 2018 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {BookendComponentInterface} from './bookend-component-interface';
+import {addAttributesToElement} from '../../../../../src/dom';
+import {dict} from '../../../../../src/utils/object';
+import {htmlFor, htmlRefs} from '../../../../../src/static-template';
+import {isArray} from '../../../../../src/types';
+import {isProtocolValid} from '../../../../../src/url';
+import {user} from '../../../../../src/log';
+
+/**
+ * @typedef {{
+ *   url: string,
+ *   text: string,
+ * }}
+ */
+let CtaLinkArrDef;
+
+/**
+ * @typedef {{
+ *   type: string,
+ *   links: !Array<CtaLinkArrDef>,
+ * }}
+ */
+export let CtaLinkDef;
+
+/**
+ * @struct @typedef {{
+ *   linkText: !Element,
+ * }}
+ */
+let ctaLinkElsDef;
+
+/**
+ * Builder class for the call to action link component.
+ * @implements {BookendComponentInterface}
+ */
+export class CtaLinkComponent {
+  /**
+   * @param {!../bookend-component.BookendComponentDef} ctaLinksJson
+   * @override
+   * */
+  assertValidity(ctaLinksJson) {
+    user().assert(ctaLinksJson['links'] && isArray(ctaLinksJson['links']) &&
+      ctaLinksJson['links'].length > 0, 'CTA link component must be an array' +
+      ' and contain at least one link inside it.');
+
+    ctaLinksJson['links'].forEach(ctaLink => {
+      user().assert('text' in ctaLink && 'url' in ctaLink, 'All links in CTA ' +
+        'link component must contain `text` field and a `url`.');
+
+      user().assert(isProtocolValid(ctaLink['url']), 'Unsupported protocol' +
+        ` for CTA link URL ${ctaLink['url']}`);
+    });
+  }
+
+  /**
+   * @param {!../bookend-component.BookendComponentDef} ctaLinksJson
+   * @return {!CtaLinkDef}
+   * @override
+   * */
+  build(ctaLinksJson) {
+    return {
+      type: ctaLinksJson['type'],
+      links: ctaLinksJson['links'],
+    };
+  }
+
+  /**
+   * @param {!../bookend-component.BookendComponentDef} ctaLinksData
+   * @param {!Document} doc
+   * @return {!Element}
+   * @override
+   * */
+  buildTemplate(ctaLinksData, doc) {
+    const html = htmlFor(doc);
+    const containerTemplate =
+        html`
+        <div class="i-amphtml-story-bookend-cta-link-wrapper">
+        </div>`;
+
+    ctaLinksData['links'].forEach(currentLink => {
```

Comment:
Naming them `templates` is an anti-pattern since it creates an actual element.

nit: Since you're recreating the same static element, I'd suggest doing something like:

```js
let linkSeed = 
    html`<a class="i-amphtml-story-bookend-cta-link" target="_top">
      <div class="i-amphtml-story-bookend-cta-link-text" ref="linkText">
      </div>
    </a>`;

ctaLinksData['links'].forEach(link => {
   const el = linkSeed.cloneNode(/* deep */ true);
   el.setAttribute('href', link['url']);

   const {linkText} = htmlRefs(el);
   linkText.textContent = link['text'];

   container.appendChild(el);
});

linkSeed = null; // GC
```
This keeps the browser from parsing the template each time.

No need for a special type for `htmlRefs` results unless it's used throughout. Since the usage is scoped, you don't need a type for safety.

Reply:
Very interesting. Thanks for the pointers!